src/geometry/GeoTrackView.i.hh

```diff
@@ -74,54 +75,106 @@ GeoTrackView& GeoTrackView::operator=(const DetailedInitializer& init)
         pos_ = init.other.pos_;
     }
     // Set up the next state and initialize the direction
-    vgnext_.Clear();
-    next_step_ = celeritas::numeric_limits<real_type>::quiet_NaN();
     dir_       = init.dir;
+
+    this->find_next_step();
     return *this;
 }
 
 //---------------------------------------------------------------------------//
 //! Find the distance to the next geometric boundary.
-CELER_FUNCTION void GeoTrackView::find_next_step()
+CELER_FUNCTION void
+GeoTrackView::find_next_step(vecgeom::VPlacedVolume const* pplvol)
 {
-    const vecgeom::LogicalVolume* logical_vol
-        = this->volume().GetLogicalVolume();
-    CELER_ASSERT(logical_vol);
+    CELER_ASSERT(pplvol);
+
     const vecgeom::VNavigator* navigator
-        = this->volume().GetLogicalVolume()->GetNavigator();
+      = pplvol->GetLogicalVolume()->GetNavigator();
     CELER_ASSERT(navigator);
 
-    next_step_
-        = navigator->ComputeStepAndPropagatedState(detail::to_vector(pos_),
-                                                   detail::to_vector(dir_),
-                                                   vecgeom::kInfLength,
-                                                   vgstate_,
-                                                   vgnext_);
+    if(this->is_outside()) {
+      // handling points outside of world volume
+      real_type step = 1.e+10;
+      next_step_     = pplvol->DistanceToIn(detail::to_vector(pos_),
+					    detail::to_vector(dir_),
+					    step);
+      vgnext_.Clear();
+      if (next_step_ < step) {
+	vgnext_.Push(pplvol);
+      }
+    }
+    else {
+      next_step_
+          = navigator->ComputeStepAndPropagatedState(detail::to_vector(pos_),
+                                                     detail::to_vector(dir_),
+                                                     vecgeom::kInfLength,
+                                                     vgstate_,
+                                                     vgnext_);
+    }
+    dirty_ = false;
 }
 
 //---------------------------------------------------------------------------//
 //! Move to the next boundary and update volume accordingly
-CELER_FUNCTION void GeoTrackView::move_next_step()
+CELER_FUNCTION real_type GeoTrackView::move_to_boundary()
 {
+    if (dirty_)
+        find_next_step();
```

Comment:
Use `this->` for member function calls to avoid ambiguity.  There are a couple of other places that need updating.